### ISSUE TITLE
File Transfer Not working - OMEMO

### ISSUE THREAD

**Issue by john4526** [ChatSecure/ChatSecure-iOS#810]
After ChatSecure version 4.1 was pushed through the app store  my images cannot will send.

I see that it is now trying to send via OMEMO vs. OTR.  I don't know what is wrong.  I get either an error stating an SSL error occurred where an secure connection cannot be established or I get Chatsecure filetransfererror error 3.

I am using prosody on the server side.  Are there specific modules that are required to support this besides HTTP_upload?

**Comment by futurealecks** [comment]
Are you using https for file transfer? If so, do you have a self signed cert?

**Comment by john4526** [comment]
I am using letsencrypt for the server cert.  Do I need a cert for the http upload domain?  I have the http_upload setup in my prosody component section of my config.  Do I need https somewhere, like as a module?

**Comment by john4526** [comment]
I have confirmed I am using http_load using https.  I can see it loading when I restart.  Now the error I get is solely the Chatsecure filetransfererror error 3 which I don't know where to start with.

**Comment by chrisballinger** [comment]
Apologies, I thought I added "user friendly" error messages, they must not be showing up. One possibility is that the TLS version is too low and iOS is rejecting the connection. You should make sure your Prosody (0.9.12) and OpenSSL (1.1.0f) versions are the latest (beyond the Debian version).

**Comment by mimi89999** [comment]
When I send a file from CS on iOS to Conversations on Android, HTTP upload works fine. When I do the other way round, without OMEMO I just see s link in CS and with OMEMO I just see an empty message. I am on stable 4.1.0.

**Comment by chrisballinger** [comment]
Our link detection code isn't ideal for https:// scheme, and will require a
"filename" like `http://example.com/file.jpg`, or an image/video/audio mime
type. If you are using http:// scheme it will never show a preview because
of Apple's ATS. When using OMEMO it shouldn't be empty or blank... possibly
an unsupported format or too large?

**Comment by mimi89999** [comment]
@chrisballinger The format is jpg, so it should be supported.
The first link is `aesgcm://lebihan.pl/xmpp-upload/index.php/dfe65e99-c857-4852-ab33-9714c28b27e7/198A16C5-F5F8-421C-BA94-3D3667C87853.jpg#04dda6a1779962762560e03d9f69c16f468367f5d3011a652147ff4ba552c1bbbd653e672920111b6d9e4712f30b0ad3`. The second is `https://lebihan.pl/xmpp-upload/index.php/dc394a38-ac6f-4b5a-8b0a-def8b21d3c01/l_-qAqrvRmuGCYVJJU6ehg.jpg
URL: https://lebihan.pl/xmpp-upload/index.php/dc394a38-ac6f-4b5a-8b0a-def8b21d3c01/l_-qAqrvRmuGCYVJJU6ehg.jpg`

**Comment by john4526** [comment]
I was able to get this to work.  I realized there were some prosody modules I was missing.

**Comment by hansenandrew** [comment]
I'm having the same issue, what Prosody modules were you missing, would love to get this to work.

**Comment by mimi89999** [comment]
@hansenandrew `mod_http_upload` or `mod_http_upload_external`

**Comment by hansenandrew** [comment]
@mimi89999 thanks, I have both of them installed but it's still not working, do you need one or the other not both, is that my problem

**Comment by mimi89999** [comment]
@hansenandrew You can't have both. Have you read and understood the doc?

**Comment by hansenandrew** [comment]
@mimi89999 Thanks for you help I've kept mod_http_upload_external

**Comment by mimi89999** [comment]
@hansenandrew Have you configured Apache or Nginx for receiving and sending files? Did you install the provided PHP script?

**Comment by hansenandrew** [comment]
@mimi89999 I have edited the share.php file and installed it into a web directory, I haven't done any configurations to Apache, my wife gets the photos on her android using conversation app

**Comment by mimi89999** [comment]
@hansenandrew So it works... Be sure to properly secure it by using a very strong, randomly generated password (mine is ~30 chars) in share.php and Prosody conf, putting the storage directory outside of the webroot and setting good permissions.

**Comment by hansenandrew** [comment]
@mimi89999 Sorry still not working on my iPhone, what configurations do I need to do on Apache

**Comment by hansenandrew** [comment]
@mimi89999 Yes I have done both the settings within share.php and withing the prosody config file

**Comment by mimi89999** [comment]
@hansenandrew It is a bit off topic here. Best place to get support with Prosody is the support MUC: `prosody@conference.prosody.im`

**Comment by hansenandrew** [comment]
@mimi89999 Ok thanks

**Comment by chrisballinger** [comment]
@hansenandrew This isn't the place to discuss Prosody configuration issues, please only post bugs related to the ChatSecure client itself. For example, if you also can't get it to work when using Conversations for Android, chances are it's a server problem and you should be looking elsewhere.

@mimi89999 Thanks for nudging this back on topic

**Comment by mimi89999** [comment]
@chrisballinger Did you find out why could images not display in ChatSecure like in https://github.com/ChatSecure/ChatSecure-iOS/issues/810#issuecomment-313469521? Is there anything I could do to help finding out without having MacOS X with XCode?

**Comment by chrisballinger** [comment]
Other people are seeing this from images sent via Conversations so I believe there is a recent(?) crypto compatibility issue.

**Comment by Pentabyteman** [comment]
I'm recently receiving ChatSecureCore.FileTransferError 3 when sending images. It used to work fine (iPhone 6 with latest iOS). I also have the problem with pictures being links despite having a .jpg file format.

**Comment by mimi89999** [comment]
@chrisballinger Any news on this? I still can't see OMEMO encrypted images.

**Comment by chrisballinger** [comment]
It should work now. Does the server support valid https?

**Comment by mimi89999** [comment]
I do have valid HTTPS with A+ grade on ssllabs.

**Comment by chrisballinger** [comment]
Hm.. 

1. "Plaintext Only" mode unencrypted transfer works fine? 
2. OMEMO text chat works?
3. Other side is Conversations or ChatSecure?

**Comment by mimi89999** [comment]
1. Yes
2. Yes
3. Conversations or Gajm. Same result.

**Comment by skumarm** [comment]
I am using ejabbered server. 
Following is my findings...

Android : Conversations
iOS : ChatSecure-iOS

Unencrypted mode of file transfer works in all the scenarios ( iPhone <--> iPhone, iPhone <--> Android )
OMEMO Encrypted mode of file transfer works in the scenarios ( iPhone <--> iPhone, iPhone --> Android ). But in **Android --> iPhone** not works.

I found **Aes Gcm** used for file's encryption and decryption. Is this make any problem with this ?

Please help me...

**Comment by licaon-kter** [comment]
@skumarm update Conversations to 2.3.5

**Comment by skumarm** [comment]
> @skumarm update Conversations to 2.3.5

So you are saying, the issue is with Conversations ( Not ChatSecure ?? )

**Comment by therob84** [comment]
After this seems to be solved - what could be the reason, that I see exactly the same behaviour between ChatSecure 5.0.2 and Conversations 2.8.2?

1. "Plaintext Only" mode unencrypted transfer works fine? YES, CS<>Conversations
2. OMEMO text chat works? YES, CS<>Conversations
3. OMEMO Image transfer works? No, only one way CS->Conversatons works, not other way.

Both on same server (jabber.de)
It happens only with one CS-contact, but works with another (other iPhone model etc..), I am on Conversations.

It had worked some weeks/months before, with other version combinations of both clients (don't know which).

Any idea?

**Comment by chrisballinger** [comment]
@therob84 Is the device using iOS 12 or iOS 13? There were some changes regarding AES GCM IV length in the latest update.

**Comment by therob84** [comment]
@chrisballinger 
The affected device is running ChatSecure 5.0.2 under iOS 13.3.1. 
(Counterpart on my side is Conversations 2.8.2+fcr on LOS 16.0/Android 9 + gajim 1.1.3)

**BUT**, solution seems to be another: 
After I re-advised my CS-contact to verify, that **all my OMEMO-devices/fingerprints are selected as "trusted/verified"** (which was not for all) - now it works! But he also reported, that it seems after some minutes that the state of the activation/trusting-butons are not saved properly.

But what puzzles me more: why OMEMO text messages are working, if the non-trusted fingerprints should be the reason? Doesn't make any sense to me. Either none OMEMO-message should work, or all (if it is not the IV-thing), independently of text/image.
:-/
But thanks for your very fast response anyway! 
Btw: are there further big feature releases worked on (A/V?)? Cheers, Robert!
